### PR TITLE
[vector tile] Add encodeUri/decodeUri functions, fix bad layer handler's browse button not working for vector tile layers

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR implements a encodeUri/decodeUri pair of functions to the vector tile provider metatada class, as well as the vector tile layer's setDataSourcePrivate function. Among other things, this fixes the broken [ browse ] button in the bad layer handler dialog for vector tile layers:

https://user-images.githubusercontent.com/1728657/134641253-f6c79c41-022c-479e-b0b7-5bbb9661d3f2.mp4

In addition, alongside @nyalldawson 's recent addition of an hyper simple vector tile data provider, it greatly improves handling of data source via python scripts. E.g., plugins which would iterate over a list of map layers and fetch file-based datasets can now easily capture file-based vector tile layers. :partying_face:

### CODE REVIEW

**Comment by nirvn** [comment]
(Unrelated test failure)

File: src/core/vectortile/qgsvectortilelayer.cpp [hunk 1]
```diff
@@ -119,6 +127,12 @@ bool QgsVectorTileLayer::loadDataSource()
   }
 
   setCrs( QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:3857" ) ) );
+
+  const QgsDataProvider::ProviderOptions providerOptions;
```
Comment (nyalldawson) [inline src/core/vectortile/qgsvectortilelayer.cpp:131]: You should set the transform context here to match the layer's transform context

File: src/core/vectortile/qgsvectortileprovidermetadata.cpp [hunk 2]
```diff
@@ -55,4 +55,35 @@ void QgsVectorTileProviderMetadata::saveConnection( const QgsAbstractProviderCon
...
+  uriComponents.insert( QStringLiteral( "type" ), dsUri.param( QStringLiteral( "type" ) ) );
+  if ( uriComponents[ QStringLiteral( "type" ) ] == QStringLiteral( "mbtiles" ) )
+  {
+    uriComponents.insert( QStringLiteral( "path" ), dsUri.param( QStringLiteral( "url" ) ) );
+  }
+  else
+  {
+    uriComponents.insert( QStringLiteral( "url" ), dsUri.param( QStringLiteral( "url" ) ) );
+  }
+  return uriComponents;
```
Comment (nyalldawson) [inline src/core/vectortile/qgsvectortileprovidermetadata.cpp:78]: There's some other Uri components used which are missing here, breaking the lossless round trip. Eg zmin/zmax, serviceType, styleUrl, referer and the auth config.

File: src/core/vectortile/qgsvectortileprovidermetadata.cpp [hunk 3]
```diff
@@ -55,4 +55,35 @@ void QgsVectorTileProviderMetadata::saveConnection( const QgsAbstractProviderCon
...
+
+QVariantMap QgsVectorTileProviderMetadata::decodeUri( const QString &uri ) const
+{
+  QgsDataSourceUri dsUri;
+  dsUri.setEncodedUri( uri );
+
+  QVariantMap uriComponents;
+  uriComponents.insert( QStringLiteral( "type" ), dsUri.param( QStringLiteral( "type" ) ) );
+  if ( uriComponents[ QStringLiteral( "type" ) ] == QStringLiteral( "mbtiles" ) )
+  {
```
Comment (nyalldawson) [inline src/core/vectortile/qgsvectortileprovidermetadata.cpp:71]: Local XYZ paths should also switch to "path" key if you want these paths to be fixable

File: src/core/vectortile/qgsvectortileprovidermetadata.cpp (hunk 2 above)
Comment (nirvn) [reply src/core/vectortile/qgsvectortileprovidermetadata.cpp:78]: @nyalldawson , regarding auth config, I'm not sure we should handle it there (ATM, we don't in any of the providers which have implemented decodeUri/encodeUri).

**Comment by nirvn** [comment]
@nyalldawson , comments addressed.

**Comment by nirvn** [comment]
Thanks to the presence of a data provider and decodeUri function, we can remove some metadata code in the vector tile layer class and instead rely on the map layer function used by all other layer types. 

image

Beyond less code to maintain, bonuses include: an hyperlink to open the folder of mbtiles, a file size, and latest modification date details.

File: src/core/vectortile/qgsvectortileprovidermetadata.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/vectortile/qgsvectortileprovidermetadata.cpp:94]: Actually they do: check afs, ams, hana, oracle, postgres.... 

If it's not handled then the lossless round trip contract is broken, which will break things

Comment (nirvn) [reply src/core/vectortile/qgsvectortileprovidermetadata.cpp:94]: oh I missed those. we re missing it for ogr and gdal then. I ll adjust.

Comment (nirvn) [reply src/core/vectortile/qgsvectortileprovidermetadata.cpp:94]: For the record I think it's also missing for the wms provider.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**